README/README.CXXMODULES.md

```diff
@@ -0,0 +1,486 @@
+of having some non-trivial patches in clang to merge default function arguments
+and default template arguments. The design of the the ROOTMAP infrastructure
+requires the default arguments to be attached to more than one declaration which
+is not allowed by standard C++. The behavior of line #1 is equivalent to:
+```cpp
+// ROOT prompt
+root [] namespace foo { };struct S;
+root [] S *s;
+```
+
+Line #2 does not require a definition, however, the second identifier lookup
+fails. The implementation knows that `foo::bar` is in *libFoo*. It `dlopen`s
+libFoo which in turn, during its static initialization, inserts annotated forward
+declaration as shown in `G__Foo.cxx`. In turn, this resolves `foo::bar` and
+parsing of `Foo.h` is again avoided at relatively small overhead. However, this
+is very hard to measure because the dictionary of each library can have different
+amount of content. In the case where the library is big and the annotated
+forward declarations are many, and we want to include a relatively small header
+file it may not pay off. Moreover, the loading of the annotated forward
+declarations can happen at any time during parsing. This is nick-named
+"recursive parsing" and is a code path that exists only in ROOT, never exercised
+by clang itself and is thus not well tested. The behavior of line #2 is
+equivalent to:
+```cpp
+// ROOT prompt
+root [] namespace foo { };struct S;
+root [] foo::bar/*store parsing state*/
+        gSystem->Load("Foo");
+        // More scaffolding.
+        extern int __Cling_Autoloading_Map;
+        namespace foo{struct __attribute__((annotate("$clingAutoload$Foo.h"))) bar;}
+        struct __attribute__((annotate("$clingAutoload$Foo.h"))) S;
+        // More initialization scaffolding.
+        /*restore parsing state*/ *baz1;
+```
+
+Line #3 requires a definition and the implementation behaves exactly as in #2.
+Then it is informed that a definition is required, it reads the information in
+the annotation and parses `Foo.h`. The recursive parsing happens at two places
+making this code path error prone.
+```cpp
+// ROOT prompt
+root [] namespace foo { };struct S;
+root [] foo::bar/*store parsing state*/
+        gSystem->Load("Foo");
+        // More scaffolding.
+        extern int __Cling_Autoloading_Map;
+        namespace foo{struct __attribute__((annotate("$clingAutoload$Foo.h"))) bar;}
+        struct __attribute__((annotate("$clingAutoload$Foo.h"))) S;
+        // More initialization scaffolding.
+        /*restore parsing state*/ baz1 /*store parsing state*/
+        #include <Foo.h>/*restore parsing state*/;
+```
+
+To recap, unfortunately, ROOT PCH is not extendable; ROOTMAP require a lot of
+maintenance and goes on a very untested codepath, while RDICT has a very limited
+scope. The three features require a lot of mechanisms to work together and the
+corner cases are very many. The interaction between some of the features are
```

Comment:
How is the interaction not well defined? Reality contradicts, also there's no need to bash the present to show improvements.

Reply:
For example, the rootmap files require reading of forward decls at startup, then loading of the library with annotated forward decls then possibly loading definitions. This require (a) a bunch of callbacks to work together; (b) patches in the compiler introducing non-standard C++; (c) demangling symbol information to guess which is the declaration; (d) guessing the header file from the name of the identifier. That’s for rootmaps and it is pretty ugly already. Rdicts and rootmap code interact mostly in TCling::RegisterModule which is pretty hard to modify and maintain in the moment. Rdict and pch interact in rootcling, where we need to keep track of what we put in the pch and write an empty rdict.

The interacting features are breaking design and introducing layering violations.

Reply:
Reworded to " The interaction between some of the features often
break design and introduce layering violations."